controllers/reaper_controller_test.go

```diff
@@ -149,35 +149,60 @@ var _ = Describe("Reaper controller", func() {
 
 		By("check that the deployment is created")
 		deploymentKey := types.NamespacedName{Namespace: ReaperNamespace, Name: ReaperName}
-		deployment := &appsv1.Deployment{}
+		currentDeployment := &appsv1.Deployment{}
 
 		Eventually(func() error {
-			return k8sClient.Get(context.Background(), deploymentKey, deployment)
+			return k8sClient.Get(context.Background(), deploymentKey, currentDeployment)
 		}, timeout, interval).Should(Succeed(), "deployment creation check failed")
 
-		Expect(len(deployment.OwnerReferences)).Should(Equal(1), "deployment owner reference not set")
-		Expect(deployment.OwnerReferences[0].UID).Should(Equal(reaper.GetUID()), "deployment owner reference has wrong uid")
+		Expect(len(currentDeployment.OwnerReferences)).Should(Equal(1), "deployment owner reference not set")
+		Expect(currentDeployment.OwnerReferences[0].UID).Should(Equal(reaper.GetUID()), "deployment owner reference has wrong uid")
 
 		By("update deployment to be ready")
-		patchDeploymentStatus(deployment, 1, 1)
+		patchDeploymentStatus(currentDeployment, 1, 1)
 
 		verifyReaperReady(types.NamespacedName{Namespace: ReaperNamespace, Name: ReaperName})
 
+		// Simulate changing the deployment strategy to rolling.
+		// The deployment strategy should be reconciled back to recreate.
+		By("patch deployment to have a rolling update strategy")
+		Eventually(func() error {
+			return k8sClient.Get(context.Background(), deploymentKey, currentDeployment)
+		}, timeout, interval).Should(Succeed(), "could not obtain current deployment")
+		deploymentPatch := client.MergeFrom(currentDeployment.DeepCopy())
+		currentDeployment.Spec.Strategy = appsv1.DeploymentStrategy{
+			Type: appsv1.RollingUpdateDeploymentStrategyType,
+		}
+		if err := k8sClient.Patch(context.Background(), currentDeployment, deploymentPatch); err != nil {
+			Fail("Failed to patch deployment with appsv1.RecreateDeploymentStrategyType")
+		}
+		Eventually(func() bool {
+			if err := k8sClient.Get(context.Background(), deploymentKey, currentDeployment); err != nil {
+				Fail("lost the reaper deployment")
```

Comment:
Do you really want to fail inside Eventually? Is there a reason for it? Wouldn't it make more sense to just exit this with false and then the Eventually block fails if all the request attempts fail?

Reply:
Yeah, we shouldn't be losing the Reaper deployment here. It has been created, if it disappears then something very strange has happened in the background. If we did somehow lose it I'd rather fail fast so we're not sitting there waiting for the timeout.